history/history.c

```diff
@@ -0,0 +1,1079 @@
+    goto error; \
+}
+
+#define check_db_rc(db, rc) if((rc) != SQLITE_OK) { \
+    fprintf(stderr, \
+        "check_db_rc failed with %s in %s line %d\n", \
+        sqlite3_errmsg(db), __FUNCTION__, __LINE__); \
+    ((void)(rc)); /* suppress "set but not used" warning */ \
+    goto error; \
+}
+
+#define check_db_cmd(db, cmd) { \
+    rc = (cmd); \
+    check_db_rc(db, rc); \
+}
+
+#define safe_free(ptr) { if ((ptr) != NULL) { free(ptr); ptr = NULL; }}
+
+#define map_isset(map, idx) (map[(idx)-1] != 0)
+#define map_set(map, idx) { map[(idx)-1] = 1;}
+#define map_unset(map, idx) { map[(idx)-1] = 0;}
+
+static
+int _cmp_int(const void *p1, const void *p2)
+{
+    return *(int *)p1 - *(int *)p2;
+}
+
+static
+void sort_array(int *arr, int count)
+{
+    qsort(arr, count, sizeof(int), _cmp_int);
+}
+
+/*
+ * Diff arr1 and arr2, with sizes count1 and count2. The arrays must be
+ * sorted. The differences will be stored in ponly1 and ponly2, which are
+ * dynamically allocated and should be free'd by the caller. Counts of the
+ * created arrays are stored in ponly1_count and ponly2_count.
+ */
+static
+int diff_arrays(int *arr1, int count1,
+                 int *arr2, int count2,
+                 int **ponly1, int *ponly1_count,
+                 int **ponly2, int *ponly2_count)
+{
+    int rc = 0;
+    int i1 = 0, i2 = 0;
+    int *only1 = NULL, *only2 = NULL;
+    int only1_count = 0, only2_count = 0;
+
+    check_ptr(only1 = (int *)calloc(count1, sizeof(int)));
+    check_ptr(only2 = (int *)calloc(count2, sizeof(int)));
+
+    while(i1 < count1 && i2 < count2) {
+        if (arr1[i1] != arr2[i2]) {
+            if(arr1[i1] > arr2[i2]) {
+                only2[only2_count++] = arr2[i2++];
+            } else {
+                only1[only1_count++] = arr1[i1++];
+            }
+        } else {
+            i1++;
+            i2++;
+        }
+    }
+
+    if(i1 == count1) {
+        while (i2 < count2) {
+            only2[only2_count++] = arr2[i2++];
+        }
+    } else {
+        while (i1 < count1) {
+            only1[only1_count++] = arr1[i1++];
+        }
+    }
+    *ponly1 = only1;
+    *ponly1_count = only1_count;
+
+    *ponly2 = only2;
+    *ponly2_count = only2_count;
+
+    return 0;
+error:
+    safe_free(only1);
+    safe_free(only2);
+    return rc;
+}
+
+static
+sqlite3 *init_db(const char *filename)
+{
+    sqlite3 *db = NULL;
+
+    int rc = sqlite3_open(filename, &db);
+    check_db_rc(db, rc);
+
+error:
+    return db;
+}
+
+/* count known rpms in db */
+static
+int db_rpms_count(sqlite3 *db, int *pcount)
+{
+    int rc = 0, step;
+    sqlite3_stmt *res = NULL;
+    const char *sql = "SELECT count(*) FROM rpms;";
+
+    rc = sqlite3_prepare_v2(db,
+                            sql,
+                            -1, &res, 0);
+    check_db_rc(db, rc);
+
+    step = sqlite3_step(res);
+    if (step == SQLITE_ROW) {
+        *pcount = sqlite3_column_int(res, 0);
+    }
+
+error:
+    if (res)
+        sqlite3_finalize(res);
```

Comment:
Can we put this in cleanup as we are releasing resource after finalizing statement

Reply:
Again, different style. There is no `cleanup` label.